ot/gromov.py

```diff
@@ -0,0 +1,478 @@
+
+    Returns
+    ----------
+    C updated
+
+
+    """
+    tmpsum = np.sum([lambdas[s] * np.dot(T[s].T, Cs[s]).dot(T[s])
+                     for s in range(len(T))])
+    ppt = np.dot(p, p.T)
+
+    return(np.divide(tmpsum, ppt))
+
+
+def update_kl_loss(p, lambdas, T, Cs):
+    """
+    Updates C according to the KL Loss kernel with the S Ts couplings calculated at each iteration
+
+
+    Parameters
+    ----------
+    p  : np.ndarray(N,)
+         weights in the targeted barycenter
+    lambdas : list of the S spaces' weights
+    T : list of S np.ndarray(ns,N)
+        the S Ts couplings calculated at each iteration
+    Cs : Cs : list of S np.ndarray(ns,ns)
+         Metric cost matrices
+
+    Returns
+    ----------
+    C updated
+
+
+    """
+    tmpsum = np.sum([lambdas[s] * np.dot(T[s].T, Cs[s]).dot(T[s])
+                     for s in range(len(T))])
+    ppt = np.dot(p, p.T)
+
+    return(np.exp(np.divide(tmpsum, ppt)))
+
+
+def gromov_wasserstein(C1, C2, p, q, loss_fun, epsilon, numItermax=1000, stopThr=1e-9, verbose=False, log=False):
+    """
+    Returns the gromov-wasserstein coupling between the two measured similarity matrices
+
+    (C1,p) and (C2,q)
+
+    The function solves the following optimization problem:
+
+    .. math::
+        \GW = arg\min_T \sum_{i,j,k,l} L(C1_{i,k},C2_{j,l})*T_{i,j}*T_{k,l}-\epsilon(H(T))
+
+        s.t. \GW 1 = p
+
+             \GW^T 1= q
+
+             \GW\geq 0
+
+    Where :
+
+        C1 : Metric cost matrix in the source space
+        C2 : Metric cost matrix in the target space
+        p  : distribution in the source space
+        q  : distribution in the target space
+        L  : loss function to account for the misfit between the similarity matrices
+        H  : entropy
+
+
+    Parameters
+    ----------
+    C1 : np.ndarray(ns,ns)
+         Metric cost matrix in the source space
+    C2 : np.ndarray(nt,nt)
+         Metric costfr matrix in the target space
+    p :  np.ndarray(ns,)
+         distribution in the source space
+    q :  np.ndarray(nt)
+         distribution in the target space
+    loss_fun :  loss function used for the solver either 'square_loss' or 'kl_loss'
+    epsilon : float
+        Regularization term >0
+    numItermax : int, optional
+        Max number of iterations
+    stopThr : float, optional
+        Stop threshold on error (>0)
+    verbose : bool, optional
+        Print information along iterations
+    log : bool, optional
+        record log if True
+    forcing : np.ndarray(N,2)
+        list of forced couplings (where N is the number of forcing)
+
+    Returns
+    -------
+    T : coupling between the two spaces that minimizes :
+            \sum_{i,j,k,l} L(C1_{i,k},C2_{j,l})*T_{i,j}*T_{k,l}-\epsilon(H(T))
+
+    """
+
+    C1 = np.asarray(C1, dtype=np.float64)
+    C2 = np.asarray(C2, dtype=np.float64)
+
+    T = np.dot(p, q.T)  # Initialization
+
+    cpt = 0
+    err = 1
+
+    while (err > stopThr and cpt < numItermax):

```

Comment:
avoid while loops. Use for with break. It's much safer to avoid infinite loops

you can use for else syntax to capture the absence of a break

Reply:
Ok for this one I will keep the consistency with the rest of the optimization method